### ISSUE TITLE
hideAfterFirstScroll leads to distorted full document screenshots

### ISSUE THREAD

**Issue by anto-ac** [wswebcreation/webdriver-image-comparison#21]
**Environment (please complete the following information):**
 - Node.js version: 8.11.4 (but also tries in 10.15.0)
 - Browser name and version: Chrome 75.0.3770.80
 - Platform name and version: Debian Jessie
 - Framework name and version: wdio 5.10.0
 - Plugin name and version: wdio-image-comparison-service 1.1.0
 - webdriver-image-comparison version: 0.5.0

**Config of the automation framework + plugin**
See here: https://github.com/anto-ac/hideAfterFirstScroll-image-comparison-service-issue

**Describe the bug**
When taking  full document screenshot using `hideAfterFirstScroll`, the generated screenshot has some issues/distortions. 

See image (some of the problems are highlighted in black circles):
https://user-images.githubusercontent.com/30859172/59304124-8d478f80-8c8f-11e9-981a-e5b881acacea.png

**To Reproduce**
Run the test included in this repo:
https://github.com/anto-ac/hideAfterFirstScroll-image-comparison-service-issue

You will need docker and docker-compose to run it.

**Expected behavior**
The whole page should be captured without distortions.

**Log**
See here: 
https://gist.github.com/anto-ac/0bdf4d51a7049a1a9c9170542cd674bb

**Additional context**

**Comment by wswebcreation** [comment]
@anto-ac 

I've been checking this on my own machine (without docker, just a Mac with the latest version of Chrome) and on a Windows machine. The strange thing is that I'm able to get a correct screenshot on 2 different screen sizes. It doesn't matter if I remove or keep the cookiebanner.

Did you run this in headless mode?

**Chrome Mac without the cookiebar:**
![treatwell-chrome-1200x983](https://user-images.githubusercontent.com/11979740/59317213-8cd6e500-8cdf-11e9-9e04-6fcad2b2896b.png)

**Chrome Windows with cookiebar:**
![treatwell-googlechrome-1366x768](https://user-images.githubusercontent.com/11979740/59317569-25ba3000-8ce1-11e9-9ddd-ba9118549af6.png)


**Chrome Mac resized without cookie bar:**
![treatwell-small-chrome-519x800](https://user-images.githubusercontent.com/11979740/59317290-d6bfcb00-8cdf-11e9-9012-a2ae51ed78af.jpg)

**Comment by anto-ac** [comment]
@wswebcreation I've also tried running the test locally on Chrome on my Mac with `--window-size=519,800` and the screenshot looks fine. 

However, it does not work as expected when running in docker as per the example in the repo I've created. The difference is that Chrome will be running in linux as a dockerised hub node.

**Comment by anto-ac** [comment]
I've also run it on a local instance of Chrome on a Linux box and the issue happens. So it seems it may be linux specific, which is interesting.

**Comment by wswebcreation** [comment]
@anto-ac 

Hmm, problem is that I don't have a Linux machine to try to check it. It's working on Windows and on a Mac, is Linux a mandatory system for your customer to check if it looks good?

I need to dive into it, but the problem is that the usage of Linux in comparison to Windows / Mac is so low that it doesn't have a high priority for me currently (if I look at all the think I also need to do)

If you like to debug it you're free to do it.

**Comment by anto-ac** [comment]
While it’s true that usage on Linux is low, the point is that anyone who is using a dockerised grid will be running the tests in Linux. And this is a much higher number than actual Linux users.

**Comment by wswebcreation** [comment]
You are right, but only for automation, and not for customer usage. And you need to be aware of the fact that for visual you need to verify the platform / browser combination that the user is using.

Same as you are doing image comparison on headless, what's the use of that, I've never seen a customer using a headless browser :wink:

So that's the reason why I put it to low priority

**Comment by anto-ac** [comment]
While doing visual regression on browsers representative of users’ statistics may be preferable, I would say that in order to identify if you’ve made a breaking change to the markup or the CSS, which is indeed the main point of visual regression, it doesn’t really matter if Chrome (for example) is running on Linux or Mac. Don’t you think?
By the way, we’re not running the tests headless as the example I put together shows.
A lot of organisations out there don’t have a grid of Macs or window machines that they can use for automation purposes, nor do they necessarily use clouds based services like browserstack.

**Comment by wswebcreation** [comment]
> Chrome (for example) is running on Linux or Mac. Don’t you think?

There is a diff in font rendering,  handling native elements like the checkbox/selectbox and so on. They all differ per platform. Even the support for certain CSS properties differ.

> A lot of organisations out there don’t have a grid of Macs or window machines that they can use for automation purposes, nor do they necessarily use clouds based services like browserstack

I understand, but then, when comparing images and validate you don't have a regression for your end users you should be testing the combinations your customers are using, that's why I mentioned the headless as an example. It's not something a user is using

**Comment by anto-ac** [comment]
I understand, but if I want to know if my CSS refactoring has introduced a regression issue, in the vast majority of cases I will find out whichever OS I’m running the browser in. 

Any kind of testing involves proxying in behalf of a user, therefore all tests are imperfect in this sense. 

My point is that, regardless of whether taking screenshot on a browser that is more representative of what a real user may be using, it doesn’t mean that not doing so has no value, especially when you take into account the necessary trade offs. By having a problem in Linux, a large base of organisations will be unable to use this library, which is otherwise amazing and that was the main reason why I’ve decided to start migrating my client’s tests to wdio v5 (the “old” visual regression service we were using is not compatible with v5)

**Comment by wswebcreation** [comment]
What do you mean with unable to use it, do all the commands fail on Linux? Or only the full page command?

**Comment by anto-ac** [comment]
I mean that the inability of having reliable full page screenshots in Linux would be a dealbreaker (for my client and I imagine for other people too)

**Comment by anto-ac** [comment]
Hi @wswebcreation!

Yesterday I installed ubuntu on virtualbox because I wanted to see the browser in action. I’ve made two videos of the full page screenshot being taken on both mac and linux, and I think I can see a difference in behaviour between the two, which might be what’s causing the problem with the final screenshot. Can you see it?

Linux
https://youtu.be/g46I15WJE50

Mac
https://youtu.be/kolOrVTIcww

I haven't had time to truly debug through the code yet (but I've got it all set up).

**Comment by wswebcreation** [comment]
Hi @anto-ac 

thanks for the movies, if you mean the flickering, that is the scrollbar that I'm hiding between screenshots. I need to have it enabled when I scroll. I doesn't show on the mac because they handle scrollbars differently.

I also need to dive into this and debug it, but I can't find the time currently

**Comment by anto-ac** [comment]
Ah yes, that's what I meant. In the Linux example though, after the scrollbar is hidden the document is resized to a larger width. So I wonder if that could be problem. 

Also, the mac one seems to scroll a few pixels less for each page down.

By the way the value of `scrollHeight` [here](https://github.com/wswebcreation/webdriver-image-comparison/blob/master/lib/methods/screenshots.ts#L284) is different for the two browsers. This is where I got so far!

**Comment by wswebcreation** [comment]
> Ah yes, that's what I meant. In the Linux example though, after the scrollbar is hidden the document is resized to a larger width. So I wonder if that could be problem.

The recalculation is done after the scrollbar is hidden, so that should not be a problem, the scrolling is only done when the page get's smaller, that might be the problem.

Thanks for trying to debug this

**Comment by anto-ac** [comment]
I’d be happy to create a PR myself, but it sounded like you wanted to think about the thing more holistically. My current thinking is:

Would you be absolutely against honouring hideScrollBars when taking a full page screenshot?

Another option would be to to never hide scroll bars when taking screenshots depending on the platform.

**Comment by wswebcreation** [comment]
Hi @anto-ac 
I've released version `1.2.0` of `wdio-image-comparison-service`, see the release notes [here](https://github.com/wswebcreation/wdio-image-comparison-service/releases/tag/v1.2.0).

Can you please verify with that version if it is fixed and if so, please close the issue.

Thanks for your help on this

**Comment by anto-ac** [comment]
Thanks @wswebcreation. Just to confirm, now the current behaviour is that the scrollbars will be hidden only before starting taking the sequence of screenshots (and displayed again afterwords) for the full page if `hideScrollBars` is true, but it will **not** be hidden and re-displayed in between screenshots.

Could you confirm please?

**Comment by wswebcreation** [comment]
@anto-ac 

Yes, you're right

**Comment by anto-ac** [comment]
It looks good! Thank you very much for this!

**Comment by wswebcreation** [comment]
Awesome, tnx for the feedback
